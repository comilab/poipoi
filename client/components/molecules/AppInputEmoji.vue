<template lang="pug">
.flex.flex-col
  AppLabel(
    v-if="label",
    data-testid="label"
  ) {{ label }}
  .mb-2
    AppTwemojiPicker(
      :disabled="disabled",
      @select="onSelect",
      data-testid="twemoji-picker"
    )
      slot(name="button")
  .w-full.border.rounded.px-2(data-testid="emojis")
    .text-gray-700.m-2(v-if="!value.length") 全ての絵文字を使用
    .emoji-list.flex.flex-wrap.overflow-auto(v-else)
      AppEmoji.w-5.h-5.m-2(
        v-for="emoji in value",
        :key="emoji",
        :text="emoji",
        :class="{ 'cursor-pointer': !disabled }",
        @click.native="remove(emoji)",
        data-testid="emoji"
      )
</template>

<script lang="ts">
import { defineComponent } from 'nuxt-composition-api'
import { PropType } from '@vue/composition-api'
import uniq from 'lodash/uniq'
import { faSmile } from '@fortawesome/free-solid-svg-icons'

import AppLabel from '~/components/atoms/AppLabel.vue'
import AppTwemojiPicker from '~/components/atoms/AppTwemojiPicker.vue'
import AppButton from '~/components/atoms/AppButton.vue'
import AppEmoji from '~/components/atoms/AppEmoji.vue'

export default defineComponent({
  components: {
    AppTwemojiPicker,
    AppLabel,
    AppButton,
    AppEmoji
  },
  props: {
    value: {
      type: Array as PropType<string[]>,
      required: true
    },
    label: {
      type: String,
      required: false,
      default: null
    },
    disabled: {
      type: Boolean,
      required: false,
      default: false
    }
  },
  setup (props, { emit }) {
    const onSelect = (emoji: string) => {
      props.value.push(emoji)
      emit('input', uniq(props.value))
    }

    const remove = (emoji: string) => {
      if (!props.disabled) {
        emit('input', props.value.filter(char => char !== emoji))
      }
    }

    return {
      onSelect,
      remove,
      faSmile
    }
  }
})
</script>

<style lang="postcss" scoped>
.emoji-list {
  min-height: 1rem;
  max-height: 10rem;
}
</style>
