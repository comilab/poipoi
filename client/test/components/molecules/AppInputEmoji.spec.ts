import { render, RenderOptions, within, fireEvent } from '@testing-library/vue'
import { defineComponent } from 'nuxt-composition-api'

import AppInputEmoji from '~/components/molecules/AppInputEmoji.vue'

describe('components/molecules/AppInputEmoji', () => {
  let options: RenderOptions<AppInputEmoji>
  let props: any

  beforeEach(() => {
    props = {
      value: []
    }

    options = {
      props
    }
  })

  test('renders correctly', () => {
    const { container } = render(AppInputEmoji, options)

    expect(container).toMatchSnapshot()
  })

  describe('props', () => {
    describe('value', () => {
      test('AppEmoji„Å´Ê∏°„Åô', () => {
        let emojiProps: any
        options.stubs = {
          AppEmoji: defineComponent({
            props: ['text'],
            setup (props) {
              emojiProps = props
            },
            template: '<span />'
          })
        }
        props.value = ['üç£']
        const { queryAllByTestId } = render(AppInputEmoji, options)

        expect(queryAllByTestId('emoji')).toHaveLength(1)
        expect(emojiProps.text).toBe('üç£')
      })

      test('Á©∫ÈÖçÂàó„ÅÆÂ†¥Âêà„ÅØ„ÉÜ„Ç≠„Çπ„Éà„ÇíË°®Á§∫', () => {
        props.value = []
        const { getByTestId } = render(AppInputEmoji, options)

        within(getByTestId('emojis')).getByText('ÂÖ®„Å¶„ÅÆÁµµÊñáÂ≠ó„Çí‰ΩøÁî®')
      })
    })

    describe('label', () => {
      test('„É©„Éô„É´„Å®„Åó„Å¶Ë°®Á§∫', () => {
        props.label = 'label'
        const { container, getByTestId } = render(AppInputEmoji, options)

        expect(container).toMatchSnapshot()
        within(getByTestId('label')).getByText('label')
      })

      test('Á©∫„Å™„Çâ„É©„Éô„É´Ë¶ÅÁ¥†Ëá™‰Ωì„ÇíÊèèÁîª„Åó„Å™„ÅÑ', () => {
        const { container, queryByTestId } = render(AppInputEmoji, options)

        expect(container).toMatchSnapshot()
        expect(queryByTestId('label')).not.toBeInTheDocument()
      })
    })

    describe('disabled', () => {
      test('AppTwemojiPicker„Å´Ê∏°„Åô', () => {
        let twemojiPickerProps: any
        options.stubs = {
          AppTwemojiPicker: defineComponent({
            props: ['disabled'],
            setup (props) {
              twemojiPickerProps = props
            },
            template: '<div />'
          })
        }
        props.disabled = true
        render(AppInputEmoji, options)

        expect(twemojiPickerProps.disabled).toBeTruthy()
      })

      test('AppEmoji„Çí„ÇØ„É™„ÉÉ„ÇØ„Åó„Å¶„ÇÇ‰Ωï„ÇÇËµ∑„Åì„Çâ„Å™„ÅÑ', async () => {
        props.value = ['üç£']
        props.disabled = true
        const { getByTestId, emitted } = render(AppInputEmoji, options)
        await fireEvent.click(getByTestId('emoji'))

        expect(emitted().input).toBeFalsy()
      })
    })
  })

  describe('slots', () => {
    describe('button', () => {
      test('AppTwemojiPickerÂÜÖ„Å´Ë°®Á§∫', () => {
        options.stubs = {
          AppTwemojiPicker: defineComponent({
            template: '<div><slot /></div>'
          })
        }
        options.slots = { button: 'button' }
        const { container, getByTestId } = render(AppInputEmoji, options)

        expect(container).toMatchSnapshot()
        within(getByTestId('twemoji-picker')).getByText('button')
      })
    })
  })

  describe('emit', () => {
    describe('select', () => {
      test('AppTwemojiPicker„Åßselect„Åó„ÅüÊôÇ„Å´Áô∫ÁÅ´', () => {
        let twemojiPickerFireSelectFunction: Function
        options.stubs = {
          AppTwemojiPicker: defineComponent({
            setup (_props, { emit }) {
              twemojiPickerFireSelectFunction = () => emit('select', 'üç£')
            },
            template: '<div />'
          })
        }
        const { emitted } = render(AppInputEmoji, options)
        twemojiPickerFireSelectFunction!()

        expect(emitted().input).toHaveLength(1)
        expect(emitted().input[0]).toStrictEqual([['üç£']])
      })

      test('AppEmoji„Çíclick„Åó„ÅüÊôÇ„Å´Áô∫ÁÅ´', async () => {
        props.value = ['üç£', 'üíØ']
        const { getAllByTestId, emitted } = render(AppInputEmoji, options)
        await fireEvent.click(getAllByTestId('emoji')[0])

        expect(emitted().input).toHaveLength(1)
        expect(emitted().input[0]).toStrictEqual([['üíØ']])
      })
    })
  })
})
