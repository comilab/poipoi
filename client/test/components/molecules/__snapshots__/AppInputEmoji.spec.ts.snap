// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/molecules/AppInputEmoji props label ラベルとして表示 1`] = `
<div>
  <div
    class="flex flex-col"
  >
    <label
      class="text-sm font-bold py-2"
      data-testid="label"
    >
      label
    </label>
    <div
      class="mb-2"
    >
      <div
        class="twemoji-picker"
        data-testid="twemoji-picker"
      >
        <div>
          <div
            id="popper-container"
            style="position: absolute; left: 0px; top: 0px; margin: 0px;"
          >
            <div
              data-popper-arrow=""
              id="arrow"
              style="position: absolute;"
            />
             
            <div
              id="popper-inner"
            >
              <!---->
            </div>
          </div>
           
          <div
            id="popper-button"
          >
            <!---->
          </div>
        </div>
         
        <span
          id="dummy-clickable-skin"
        />
      </div>
    </div>
    <div
      class="w-full border rounded px-2"
      data-testid="emojis"
    >
      <div
        class="text-gray-700 m-2"
      >
        全ての絵文字を使用
      </div>
    </div>
  </div>
</div>
`;

exports[`components/molecules/AppInputEmoji props label 空ならラベル要素自体を描画しない 1`] = `
<div>
  <div
    class="flex flex-col"
  >
    <!---->
    <div
      class="mb-2"
    >
      <div
        class="twemoji-picker"
        data-testid="twemoji-picker"
      >
        <div>
          <div
            id="popper-container"
            style="position: absolute; left: 0px; top: 0px; margin: 0px;"
          >
            <div
              data-popper-arrow=""
              id="arrow"
              style="position: absolute;"
            />
             
            <div
              id="popper-inner"
            >
              <!---->
            </div>
          </div>
           
          <div
            id="popper-button"
          >
            <!---->
          </div>
        </div>
         
        <span
          id="dummy-clickable-skin"
        />
      </div>
    </div>
    <div
      class="w-full border rounded px-2"
      data-testid="emojis"
    >
      <div
        class="text-gray-700 m-2"
      >
        全ての絵文字を使用
      </div>
    </div>
  </div>
</div>
`;

exports[`components/molecules/AppInputEmoji renders correctly 1`] = `
<div>
  <div
    class="flex flex-col"
  >
    <!---->
    <div
      class="mb-2"
    >
      <div
        class="twemoji-picker"
        data-testid="twemoji-picker"
      >
        <div>
          <div
            id="popper-container"
            style="position: absolute; left: 0px; top: 0px; margin: 0px;"
          >
            <div
              data-popper-arrow=""
              id="arrow"
              style="position: absolute;"
            />
             
            <div
              id="popper-inner"
            >
              <!---->
            </div>
          </div>
           
          <div
            id="popper-button"
          >
            <!---->
          </div>
        </div>
         
        <span
          id="dummy-clickable-skin"
        />
      </div>
    </div>
    <div
      class="w-full border rounded px-2"
      data-testid="emojis"
    >
      <div
        class="text-gray-700 m-2"
      >
        全ての絵文字を使用
      </div>
    </div>
  </div>
</div>
`;

exports[`components/molecules/AppInputEmoji slots button AppTwemojiPicker内に表示 1`] = `
<div>
  <div
    class="flex flex-col"
  >
    <!---->
    <div
      class="mb-2"
    >
      <div
        data-testid="twemoji-picker"
      >
        button
      </div>
    </div>
    <div
      class="w-full border rounded px-2"
      data-testid="emojis"
    >
      <div
        class="text-gray-700 m-2"
      >
        全ての絵文字を使用
      </div>
    </div>
  </div>
</div>
`;
